Plugins/PluginsShared/PluginUtils.swift

```diff
@@ -0,0 +1,204 @@
+///   - fileNaming: The file naming scheme to be used.
+///   - inputFiles: The input `.proto` files.
+///   - protoDirectoryPaths: The directories in which `protoc` will search for imports.
+///   - protocGenGRPCSwiftPath: The path to the `proto-gen-grpc-swift` `protoc` plugin.
+///   - outputDirectory: The directory in which generated source files are created.
+/// - Returns: The constructed arguments to be passed to `protoc` when invoking the `proto-gen-grpc-swift` `protoc` plugin.
+func constructProtocGenGRPCSwiftArguments(
+  config: CommonConfiguration,
+  using fileNaming: CommonConfiguration.FileNaming?,
+  inputFiles: [URL],
+  protoDirectoryPaths: [URL],
+  protocGenGRPCSwiftPath: URL,
+  outputDirectory: URL
+) -> [String] {
+  // Construct the `protoc` arguments.
+  var protocArgs = [
+    "--plugin=protoc-gen-grpc-swift=\(protocGenGRPCSwiftPath.relativePath)",
+    "--grpc-swift_out=\(outputDirectory.relativePath)",
+  ]
+
+  if let importPaths = config.importPaths {
+    for path in importPaths {
+      protocArgs.append("-I")
+      protocArgs.append("\(path)")
+    }
+  }
+
+  if let visibility = config.visibility {
+    protocArgs.append("--grpc-swift_opt=Visibility=\(visibility.rawValue.capitalized)")
+  }
+
+  if let generateServerCode = config.server {
+    protocArgs.append("--grpc-swift_opt=Server=\(generateServerCode)")
+  }
+
+  if let generateClientCode = config.client {
+    protocArgs.append("--grpc-swift_opt=Client=\(generateClientCode)")
+  }
+
+  // TODO: Don't currently support reflection data
+  //  if let generateReflectionData = config.reflectionData {
+  //    protocArgs.append("--grpc-swift_opt=ReflectionData=\(generateReflectionData)")
+  //  }
+
+  if let fileNaming = fileNaming {
+    protocArgs.append("--grpc-swift_opt=FileNaming=\(fileNaming.rawValue)")
+  }
+
+  if let protoPathModuleMappings = config.protoPathModuleMappings {
+    protocArgs.append("--grpc-swift_opt=ProtoPathModuleMappings=\(protoPathModuleMappings)")
+  }
+
+  if let useAccessLevelOnImports = config.useAccessLevelOnImports {
+    protocArgs.append("--grpc-swift_opt=UseAccessLevelOnImports=\(useAccessLevelOnImports)")
+  }
+
+  protocArgs.append(contentsOf: protoDirectoryPaths.map { "--proto_path=\($0.relativePath)" })
+
+  protocArgs.append(contentsOf: inputFiles.map { $0.relativePath })
+
+  return protocArgs
+}
+
+/// Derive the expected output file path to match the behavior of the `proto-gen-swift` and `proto-gen-grpc-swift` `protoc` plugins.
+/// - Parameters:
+///   - inputFile: The input `.proto` file.
+///   - fileNaming: The file naming scheme.
+///   - protoDirectoryPath: The root path to the source `.proto` files used as the reference for relative path naming schemes.
+///   - outputDirectory: The directory in which generated source files are created.
+///   - outputExtension: The file extension to be appended to generated files in-place of `.proto`.
+/// - Returns: The expected output file path.
+func deriveOutputFilePath(
```

Comment:
IIRC the output file path is _only_ needed by the build plugin which can only use the `.fullPath` file naming. If so we can simplify this, right?